File: src/main/java/com/fifo/ticketing/domain/user/repository/UserRepository.java
package com.fifo.ticketing.domain.user.repository;

import com.fifo.ticketing.domain.user.entity.Role;
import com.fifo.ticketing.domain.user.entity.User;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.repository.JpaRepository;

public interface UserRepository extends JpaRepository<User, Long> {

    Optional<User> findByEmail(String email);

    Page<User> findByUsernameContaining(String userName, Pageable pageable);

    Page<User> findByRole(Role role, Pageable pageable);

    Page<User> findByUsernameContainingAndRole(String username, Role role, Pageable pageable);

}

File: src/main/java/com/fifo/ticketing/domain/user/repository/UserRepository.kt
package com.fifo.ticketing.domain.user.repository

import com.fifo.ticketing.domain.user.entity.Role
import com.fifo.ticketing.domain.user.entity.User
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.JpaRepository
import java.util.*

interface UserRepository : JpaRepository<User, Long> {
    fun findByEmail(email: String): User?

    fun findByUsernameContaining(userName: String, pageable: Pageable): Page<User>

    fun findByRole(role: Role, pageable: Pageable): Page<User>

    fun findByUsernameContainingAndRole(
        username: String,
        role: Role,
        pageable: Pageable
    ): Page<User>
}
